src/transcript.rs

```diff
@@ -0,0 +1,121 @@
+use ark_ff::PrimeField;
+
+use ark_crypto_primitives::sponge::{
+    constraints::CryptographicSpongeVar,
+    poseidon::{constraints::PoseidonSpongeVar, PoseidonConfig, PoseidonSponge},
+    Absorb, CryptographicSponge,
+};
+use ark_r1cs_std::fields::fp::FpVar;
+use ark_relations::r1cs::{ConstraintSystemRef, SynthesisError};
+
+use crate::Transcript;
+
+/// PoseidonTranscript implements the Transcript trait using the Poseidon hash
+pub struct PoseidonTranscript<F: PrimeField + Absorb> {
+    sponge: PoseidonSponge<F>,
+}
+impl<F: PrimeField + Absorb> Transcript<F> for PoseidonTranscript<F> {
+    type TranscriptConfig = PoseidonConfig<F>;
+
+    fn new(poseidon_config: &Self::TranscriptConfig) -> Self {
+        let sponge = PoseidonSponge::<F>::new(poseidon_config);
+        Self { sponge }
+    }
+    fn absorb(&mut self, v: &F) {
+        self.sponge.absorb(&v);
+    }
+    fn absorb_vec(&mut self, v: &[F]) {
+        self.sponge.absorb(&v);
+    }
+    fn get_challenge(&mut self) -> F {
+        let c = self.sponge.squeeze_field_elements(1);
+        self.sponge.absorb(&c[0]);
+        c[0]
+    }
+    fn get_challenges(&mut self, n: usize) -> Vec<F> {
+        let c = self.sponge.squeeze_field_elements(n);
+        self.sponge.absorb(&c);
+        c
+    }
+}
+
+/// PoseidonTranscriptVar implements the gadget compatible with PoseidonTranscript
+pub struct PoseidonTranscriptVar<F: PrimeField> {
+    sponge: PoseidonSpongeVar<F>,
+}
+impl<F: PrimeField> PoseidonTranscriptVar<F> {
+    pub fn new(cs: ConstraintSystemRef<F>, poseidon_config: &PoseidonConfig<F>) -> Self {
+        let sponge = PoseidonSpongeVar::<F>::new(cs, poseidon_config);
+        Self { sponge }
+    }
+    pub fn absorb(&mut self, v: FpVar<F>) -> Result<(), SynthesisError> {
+        self.sponge.absorb(&v)
+    }
+    pub fn absorb_vec(&mut self, v: &[FpVar<F>]) -> Result<(), SynthesisError> {
+        self.sponge.absorb(&v)
+    }
+    pub fn get_challenge(&mut self) -> Result<FpVar<F>, SynthesisError> {
+        let c = self.sponge.squeeze_field_elements(1)?;
+        self.sponge.absorb(&c[0])?;
+        Ok(c[0].clone())
+    }
+    pub fn get_challenges(&mut self, n: usize) -> Result<Vec<FpVar<F>>, SynthesisError> {
+        let c = self.sponge.squeeze_field_elements(n)?;
+        self.sponge.absorb(&c)?;
+        Ok(c)
+    }
+}
+
+/// WARNING the method poseidon_test_config is for tests only
```

Comment:
If it's only used in tests, perhaps move it into mod `tests`?

Reply:
Right! moved into mod `tests`